src/pkgcheck/checks/codingstyle.py

```diff
@@ -903,3 +903,118 @@ def feed(self, pkg):
                         yield RedundantDodir(
                             cmd.group('cmd'), line=dodir.group('call'),
                             lineno=lineno - 1, pkg=pkg)
+
+
+class UnquotedVariable(results.AliasResult, results.Warning):
+    """Variable is used unquoted in a context where it should be quoted.
+
+    Variables like D, FILESDIR, etc may not be safe to use unquoted in some
+    contexts.
+    """
+
+    _name = 'UnquotedVariable'
+
+    def __init__(self, variable, lines, **kwargs):
+        super().__init__(**kwargs)
+        self.variable = variable
+        self.lines = tuple(lines)
+
+    @property
+    def desc(self):
+        s = pluralism(self.lines)
+        lines = ', '.join(map(str, self.lines))
+        return f'unquoted variable {self.variable} on line{s}: {lines} '
+
+
+
+class EbuildUnquotedVariable(UnquotedVariable, results.VersionResult):
+    __doc__ = UnquotedVariable.__doc__
+
+
+class EclassUnquotedVariable(UnquotedVariable, results.EclassResult):
+    __doc__ = UnquotedVariable.__doc__
+
+    @property
+    def desc(self):
+        return f'{self.eclass}: {super().desc}'
+
+
+class _UnquotedVariablesCheck(Check):
+    """Scan files for variables that should be quoted like D, FILESDIR, etc."""
+
+    message_commands = (
+        "die", "echo", "eerror", "einfo", "elog", "eqawarn", "ewarn"
+    )
+    var_names = (
+        "D", "DISTDIR", "FILESDIR", "S", "T", "ROOT", "BROOT", "WORKDIR", "ED",
+        "EPREFIX", "EROOT", "SYSROOT", "ESYSROOT", "TMPDIR", "HOME",
+        # variables for multibuild.eclass
+        "BUILD_DIR",
+    )
+
+    node_types_ok = (
+        # Variable is sitting in a string, all good
+        'string',
+        # Variable is part of a shell assignment, and does not need to be
+        # quoted. for example s=${workdir}/${pn} is ok.
+        'variable_assignment',
+        # Variable sits inside a [[ ]] test command and it's OK not to be quoted
+        'test_command',
+        # Variable is being used in a heredoc body, no need to specify quotes.
+        'heredoc_body',
+    )
+
+    def _var_needs_quotes(self, pkg, node):
+        pnode = node.parent
+        while pnode != node:
+            if pnode.type in self.node_types_ok:
+                return False
+            elif pnode.type == 'command':
+                cmd = pkg.node_str(pnode.child_by_field_name('name'))
+                return cmd not in self.message_commands
+            elif pnode.type in 'array':
+                # Variable is sitting unquoted in an array
+                return True
+            pnode = pnode.parent
+
+        # Default: The variable should be quoted
+        return True
+
+    def _feed(self, item):
+        if item.tree.root_node.has_error:
+            # Do not run this check if the parse tree contains errors, as it
+            # might result in false positives. This check appears to be quite
```

Comment:
Let's use here `item` instead of `pkg` (as it also handles eclasses now).

Reply:
You might have looked at an older commit, I changed that already.